### ISSUE TITLE
Implement the Coffman-Graham Algorithm for Scheduling

### ISSUE THREAD

**Issue by gabyx** [gabyx/ExecutionGraph#1]
# Coffman-Graham Algorithm

<div>
<ul><li>Assume all tasks execute the <em>same</em> amount of time.
<ul><li><math></math><var>mu</var>(T<sub>i</sub>) = 1.
</li></ul>

</li><li>Definitions:
<ul><li>If <math>T</math><sub>i</sub> &lt; T<sub>j</sub>, then <math>T</math><sub>j</sub> is an <dfn>immediate successor</dfn> of <math>T</math><sub>i</sub>.
</li><li>Let <math>S(T</math><sub>i</sub>) be all immediate successors of <math>T</math><sub>i</sub>.
</li><li>Let <math></math><var>alpha</var>(T) be an integer label assigned to <math>T</math>.
</li><li>Let <math>N(T</math><sub>i</sub>) be the ordered sequence of integers formed from the
set <math>{ </math><var>alpha</var>(T') | T' <var>in</var>S(T)
}.
</li></ul>

</li></ul>
</div>

<div>
<ul><li>Find: task list for Graham's algorithm.
</li><li>Algorithm:
<ol><li>Choose <math>T</math><sub>k</sub> with <math>S(T</math><sub>k</sub>)=0 and let <math></math><var>alpha</var>(T<sub>k</sub>) be 1.
</li><li>For <math>i &lt;-2</math> to <math>n</math> do
<ol><li>Let <math>R</math> be the set of unlabeled tasks with no unlabeled successors.
</li><li>Let <math>T</math><sup>*</sup> be the task in <math>R</math> such that <math>N(T</math><sup>*</sup>) is lexicographically
smaller than <math>N(T)</math> for all <math>T</math> in <math>R</math>.
</li><li>Let <math></math><var>alpha</var>(T<sup>*</sup>) &lt;-i.
</li></ol>

</li><li>Construct <math>L = (U</math><sub>n</sub>, U<sub>n-1</sub>, ..., U<sub>2</sub>, U<sub>1</sub>) such that
<math></math><var>alpha</var>(U<sub>i</sub>) = i.
</li></ol>

</li><li>Theorem: <math>w/w</math><sub>0</sub> &lt;=2 - 2/p 
<ul><li><math>w</math> length of constructed schedule, <math>w</math><sub>0</sub> length of optimal schedule,
<math>p &gt; 1</math> processor number.
</li></ul>

</li><li>Collary: schedule optimal for <math>p=2</math>.
</li></ul>
</div>

**Comment by gabyx** [comment]
<div class="section-inner sectionLayout--insetColumn"><h1 name="3df7" id="3df7" class="graf graf--h3 graf--leading graf--title">Scheduling Tasks and Drawing Graphs — The Coffman-Graham Algorithm</h1><p name="35d0" id="35d0" class="graf graf--p graf-after--h3">When an algorithm developed for one problem domain applies almost verbatim to another, completely unrelated domain, that is the type of insight, beauty and depth that makes computer science a science on its own, and not a branch of something else, namely mathematics, like many professionals educated in the field mistakenly believe. For example, one of the common algorithmic problems during the 60s was the scheduling of tasks on multiprocessor machines. The problem is, you are given a large set of tasks, some of which depend on others, that have to be scheduled for processing on N number of processors in such a way as to maximize processor use. A well-known algorithm for this problem is the <a href="https://en.wikipedia.org/wiki/Coffman%E2%80%93Graham_algorithm" data-href="https://en.wikipedia.org/wiki/Coffman%E2%80%93Graham_algorithm" class="markup--anchor markup--p-anchor" rel="nofollow noopener" target="_blank">Coffman-Graham algorithm.</a> It assumes that there are no circular dependencies between the tasks, as is usually the case when it comes to real world tasks, except in catch 22 situations at some bureaucracies run amok! To do that, the tasks and their dependencies are modeled as a DAG (a directed acyclic graph). In mathematics, this is also known as a partial order: if a tasks T1 depends on T2, we say that T2 preceeds T1, and we write T2 &lt; T1. The ordering is called <em class="markup--em markup--p-em">partial</em> because not all tasks are related in this precedence relation, some are simply independent of each other and can be safely carried out in parallel.</p><p name="1b8a" id="1b8a" class="graf graf--p graf-after--p">The Coffman-Graham algorithm works by creating a sequence of execution rounds where at each round at most N tasks execute simultaneously. The algorithm also has to make sure that all dependencies of the current round have been executed in previous rounds. Those two constraints are what makes the problem non-trivial: we want exactly N tasks at each round of execution if possible, so that all processors get used, and we also have to complete all tasks that precede a given task T before scheduling it. There are 3 basic steps to achieving the objective:</p><ol class="postList"><li name="df4e" id="df4e" class="graf graf--li graf-after--p">Cleanup the graph so that only direct dependencies are represented. So if there is a task A that depends on B and B depends on another task C, we already know that A depends indirectly on C (transitivity is one of defining features of partial orders), so that dependency does not need to be stated explicitly. Sometimes the input of a problem will have such superfluous information, but in fact this could only confuse the algorithm! Removing the indirect dependencies is called <a href="https://en.wikipedia.org/wiki/Transitive_reduction" data-href="https://en.wikipedia.org/wiki/Transitive_reduction" class="markup--anchor markup--li-anchor" rel="nofollow noopener" target="_blank">transitive reduction</a>, as opposed to the more commonly known operation of <a href="https://en.wikipedia.org/wiki/Transitive_closure" data-href="https://en.wikipedia.org/wiki/Transitive_closure" class="markup--anchor markup--li-anchor" rel="nofollow noopener" target="_blank">transitive closure</a> which explicitly computes all indirect dependencies.</li><li name="2ffb" id="2ffb" class="graf graf--li graf-after--li">Order the tasks in a single list so that the dependent ones come after their dependencies and they are sort of evenly spread apart. This is the crucial and most interesting part of the algorithm. So how are we to compare two tasks and decide which one should be run first. The trick is to proceed from the starting tasks, the roots of the graph that don’t have any dependencies whatsoever, and then progressively add tasks that depend only on them and then tasks then depend only on them etc. This is called <a href="https://en.wikipedia.org/wiki/Topological_sorting" data-href="https://en.wikipedia.org/wiki/Topological_sorting" class="markup--anchor markup--li-anchor" rel="nofollow noopener" target="_blank">topological ordering</a> of the dependency graph. There are usually many possible such orderings and some of them will lead to a good balanced distribution of tasks for the purpose of CPU scheduling while others will leave lots of CPUs unused. In step (3) of the algorithm, we are just going to take the tasks one by one from this ordering and assign them to execution rounds as they come. Therefore, to make it so that at each round, the number of CPUs is maximized, the ordering must somehow space the dependencies apart as much as possible. That is, if the order is written as [T1, T2, T3,&nbsp;…, Tn] and if Tj depends on Ti, we want j-i to be as big as possible. Intuitively, this is desirable because the closer they are, the sooner we’d have to schedule Tj for execution after Ti, and since they can’t be executed on the same parallel round, we’d end up with unused CPUs. To space the tasks apart, here is what we do. Suppose we have tasks A and B, with all their dependencies already ordered in our list and we have to pick which one is going to come next. From A’s dependencies, we take the one most recently placed in the ordering and we check if it comes before or after the most recently placed task from B’s dependencies. If it comes before, then we choose A, if it comes after then we chose B. If it turns out A and B’s most recently placed dependency is actually the same task that both depend on, we look at the next most recent dependency etc. This way, by picking the next task as the one whose closest dependency is the furthest away, at every step we space out dependencies in our ordering as much as possible.</li><li name="717a" id="717a" class="graf graf--li graf-after--li">Assign tasks to rounds so as to maximize the number of tasks executed on each round. This is the easiest step - we just reap the rewards from doing all the hard work of the previous steps. Going through our ordering [T1, T2, T3,&nbsp;…, Tn], we fill up available CPUs by assigning the tasks one by one. When all CPUs are occupied, we move to the next round and we start filling CPUs again. If while at a particular round the next task to be scheduled has a dependency that’s also scheduled for that same round, we have no choice but to leave the remaining CPUs unused and start the next round. The algorithm does not take into account how long each tasks can potentially take.</li></ol><p name="c3a9" id="c3a9" class="graf graf--p graf-after--li">Now, I said that this algorithm is also used to solve a completely different problem. The problem I was referring to is drawing networks in a visually appealing way. This is a pretty difficult problem and there are many different approaches whose effectiveness often depends on the structure of the network. When a network is devoid of cycles (paths from on node back to itself), the Coffman-Grahan algorithm just described can be applied!</p><figure name="d345" id="d345" class="graf graf--figure graf-after--p"><div class="aspectRatioPlaceholder is-locked" style="max-width: 350px; max-height: 351px;"><div class="aspectRatioPlaceholder-fill" style="padding-bottom: 100.29999999999998%;"></div><div class="progressiveMedia js-progressiveMedia graf-image is-canvasLoaded is-imageLoaded" data-image-id="1*Uao03T88sh5RvjThl24Fjg.png" data-width="350" data-height="351" data-scroll="native"><img src="https://cdn-images-1.medium.com/freeze/max/30/1*Uao03T88sh5RvjThl24Fjg.png?q=20" crossorigin="anonymous" class="progressiveMedia-thumbnail js-progressiveMedia-thumbnail"><canvas class="progressiveMedia-canvas js-progressiveMedia-canvas" width="75" height="75"></canvas><img class="progressiveMedia-image js-progressiveMedia-image" data-src="https://cdn-images-1.medium.com/max/800/1*Uao03T88sh5RvjThl24Fjg.png" src="https://cdn-images-1.medium.com/max/800/1*Uao03T88sh5RvjThl24Fjg.png"><noscript class="js-progressiveMedia-inner"></noscript></div></div></figure><p name="a6fe" id="a6fe" class="graf graf--p graf-after--figure">The idea is to think of the network nodes as the tasks and of the network connections as the dependencies between the tasks, and then build a list of consecutive layers analogous to the task execution rounds. Instead of specifying a number of available CPUs, one specifies how many nodes per layer are allowed, which is generally convenient because the drawing is done on a fixed width computer screen. Because the algorithm does not like circular dependencies, there is an extra step here to remove a select set of connections so that the network becomes a DAG. This is in addition to transitive reduction where we only keep direct connections and drop all the rest. Once the algorithm is complete, the drawing of those provisionally removed connections can be performed on top of the nice layering produced. Thus, the Coffman-Graham is (also!<strong class="markup--strong markup--p-strong">)</strong> one of <a href="https://cs.brown.edu/~rt/gdhandbook/chapters/hierarchical.pdf" data-href="https://cs.brown.edu/~rt/gdhandbook/chapters/hierarchical.pdf" class="markup--anchor markup--p-anchor" rel="nofollow noopener" target="_blank">hierarchical drawing algorithms</a>, a general framework for graph drawing developed by <a href="https://en.wikipedia.org/wiki/Kozo_Sugiyama" data-href="https://en.wikipedia.org/wiki/Kozo_Sugiyama" class="markup--anchor markup--p-anchor" rel="nofollow noopener" target="_blank">Kozo Sugiyama</a>.</p><p name="ada6" id="ada6" class="graf graf--p graf-after--p graf--trailing">The commonality between those two problems at a deeper level probably has to do with how our brains make sense of a web of dependencies. Visualizing a graph in terms of layers does give a relatively clean ordering where you can read off the relationships in stages with a limited amount of dependencies to think of in a simultaneous fashion.</p></div>
